compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_globals.rs

```diff
@@ -1,24 +1,213 @@
+use std::collections::BTreeMap;
+
 use acvm::FieldElement;
 use fxhash::{FxHashMap as HashMap, FxHashSet as HashSet};
 
-use super::{BrilligArtifact, BrilligBlock, BrilligVariable, FunctionContext, Label, ValueId};
-use crate::{
-    brillig::{brillig_ir::BrilligContext, DataFlowGraph},
-    ssa::ir::dfg::GlobalsGraph,
+use super::{
+    BrilligArtifact, BrilligBlock, BrilligVariable, Function, FunctionContext, Label, ValueId,
+};
+use crate::brillig::{
+    brillig_ir::BrilligContext, called_functions_vec, Brillig, DataFlowGraph, FunctionId,
+    Instruction, Value,
 };
 
+/// Context structure for generating Brillig globals
+/// it stores globals related data required for code generation of regular Brillig functions.
+#[derive(Default)]
+pub(crate) struct BrilligGlobals {
+    /// Both `used_globals` and `brillig_entry_points` need to be built
+    /// from a function call graph.
+    ///
+    /// Maps a Brillig function to the globals used in that function.
+    /// This includes all globals used in functions called internally.
+    used_globals: HashMap<FunctionId, HashSet<ValueId>>,
+    /// Maps a Brillig entry point to all functions called in that entry point.
+    /// This includes any nested calls as well, as we want to be able to associate
+    /// any Brillig function with the appropriate global allocations.
+    brillig_entry_points: HashMap<FunctionId, HashSet<FunctionId>>,
+
+    /// Maps an inner call to its Brillig entry point
+    /// This is simply used to simplify fetching global allocations when compiling
+    /// individual Brillig functions.
+    inner_call_to_entry_point: HashMap<FunctionId, Vec<FunctionId>>,
+    /// Final map that associated an entry point with its Brillig global allocations
+    entry_point_globals_map: HashMap<FunctionId, SsaToBrilligGlobals>,
+}
+
+/// Mapping of SSA value ids to their Brillig allocations
+pub(crate) type SsaToBrilligGlobals = HashMap<ValueId, BrilligVariable>;
+
+impl BrilligGlobals {
+    pub(crate) fn new(
+        functions: &BTreeMap<FunctionId, Function>,
+        mut used_globals: HashMap<FunctionId, HashSet<ValueId>>,
+        main_id: FunctionId,
+    ) -> Self {
+        let mut brillig_entry_points = HashMap::default();
+        for (_, function) in functions.iter() {
+            if function.runtime().is_acir() {
```

Comment:
```suggestion
        let brillig_functions = functions.iter().filter(|func| func.is_brillig());
        for (_, function) in brillig_functions {
```

I'd prefer to use iterators to do any initial preprocessing like this.